lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs

```diff
@@ -0,0 +1,531 @@
+prop_delete_size :: TxIn -> UTxOIndex -> Property
+prop_delete_size i u =
+    UTxOIndex.size (UTxOIndex.delete i u) === expected
+  where
+    expected = case UTxOIndex.lookup i u of
+        Nothing ->
+            UTxOIndex.size u
+        Just _ ->
+            UTxOIndex.size u - 1
+
+prop_insert_assets :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_assets i o u =
+    UTxOIndex.assets (UTxOIndex.insert i o u)
+        `Set.intersection` insertedAssets === insertedAssets
+  where
+    insertedAssets = TokenBundle.getAssets $ view #tokens o
+
+prop_insert_balance :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_balance i o u =
+    UTxOIndex.balance (UTxOIndex.insert i o u) === expected
+  where
+    expected = view #tokens o `TokenBundle.add` case UTxOIndex.lookup i u of
+        Nothing ->
+            UTxOIndex.balance u
+        Just o' ->
+            UTxOIndex.balance u `TokenBundle.unsafeSubtract` view #tokens o'
+
+prop_insert_delete :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_delete i o u =
+    UTxOIndex.delete i (UTxOIndex.insert i o u) === expected
+  where
+    expected =
+        if UTxOIndex.member i u then UTxOIndex.delete i u else u
+
+prop_insert_lookup :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_lookup i o u =
+    UTxOIndex.lookup i (UTxOIndex.insert i o u) === Just o
+
+prop_insert_size :: TxIn -> TxOut -> UTxOIndex -> Property
+prop_insert_size i o u =
+    UTxOIndex.size (UTxOIndex.insert i o u) === expected
+  where
+    expected = case UTxOIndex.lookup i u of
+        Nothing ->
+            UTxOIndex.size u + 1
+        Just _ ->
+            UTxOIndex.size u
+
+--------------------------------------------------------------------------------
+-- Index selection properties
+--------------------------------------------------------------------------------
+
+-- | Attempt to select a random entry from an empty index.
+--
+-- This should always return 'Nothing'.
+--
+prop_selectRandom_empty :: SelectionFilter -> Property
+prop_selectRandom_empty f = monadicIO $ do
+    result <- run $ UTxOIndex.selectRandom UTxOIndex.empty f
+    assert $ isNothing result
+
+-- | Attempt to select a random entry from a singleton index with entry 'e'.
+--
+-- This should always return 'Just e'.
+--
+prop_selectRandom_singleton :: SelectionFilter -> TxIn -> TxOut -> Property
+prop_selectRandom_singleton selectionFilter i o = monadicIO $ do
+    actual <- run $ UTxOIndex.selectRandom index selectionFilter
+    assert $ actual == expected
+  where
+    index = UTxOIndex.singleton i o
+    expected = case selectionFilter of
+        Any ->
+            Just ((i, o), UTxOIndex.empty)
+        WithAsset a | txOutHasAsset o a ->
+            Just ((i, o), UTxOIndex.empty)
+        WithAsset _ ->
+            Nothing
+
+-- | Attempt to select a random entry with any combination of assets.
+--
+-- This should always succeed, provided the index is not empty.
+--
+prop_selectRandom_one_any :: UTxOIndex -> Property
```

Comment:
Alright, this is the property test that actually tests that the implementation of the function doesn't misbehave (e.g. falsely returning Nothing).

But, it relies on other functions not being buggy (such as `delete` etc). There's a very low chance a bug may slip through. This is why I think we must have unit tests as well. Property tests are not a substitute for unit tests.